File: tests/test_scorer.py
"""Tests for the Python scorer module."""

import pytest
from backend.engine.scorer import score_username, pre_score_from_metadata


class TestScoreUsername:
    def test_bot_digits_pattern(self):
        score, details = score_username("bot8374629")
        assert score > 0

    def test_normal_username(self):
        score, details = score_username("jean_dupont")
        assert score == 0

    def test_all_digits(self):
        score, details = score_username("8374629102")
        assert score > 0

    def test_very_long_username(self):
        # 26+ chars triggers "very_long" pattern
        score, details = score_username("a" * 27)
        assert score > 0

    def test_no_vowels(self):
        score, details = score_username("bcdfghjklmnp")
        assert score > 0

    def test_high_digit_ratio(self):
        score, details = score_username("user12345678")
        assert score > 0

    def test_underscore_heavy(self):
        score, details = score_username("_a_b_c_d_e_")
        assert score > 0

    def test_capped_at_30(self):
        score, details = score_username("_bot_12345678901234567890_bcdfg_")
        assert score <= 30


class TestPreScoreFromMetadata:
    def test_obvious_fake(self):
        score, details = pre_score_from_metadata("bot8374629", 0, False, None, False)
        assert score is not None
        assert score >= 75

    def test_obvious_legit(self):
        score, details = pre_score_from_metadata("jean_dupont", 150, False, "Jean Dupont", True)
        assert score is not None
        assert score <= 15

    def test_inconclusive_returns_none(self):
        # Low followers, no pic, no name but normal username → score ~40, inconclusive
        score, details = pre_score_from_metadata("user_test", 5, False, None, False)
        assert score is None

File: tests/test_scorer.js
/**
 * Tests for the JavaScript scorer module.
 * Run with: node tests/test_scorer.js
 */

import { scoreUsername, preScoreFromMetadata, scoreProfile } from "../extension/background/scorer.js";

let passed = 0;
let failed = 0;

function assert(condition, msg) {
  if (condition) {
    passed++;
    console.log(`  PASS: ${msg}`);
  } else {
    failed++;
    console.error(`  FAIL: ${msg}`);
  }
}

function suite(name, fn) {
  console.log(`\n${name}`);
  fn();
}

// ── scoreUsername ────────────────────────────────────────────────────────────

suite("scoreUsername", () => {
  assert(scoreUsername("bot8374629") > 0, "bot+digits detected");
  assert(scoreUsername("jean_dupont") === 0, "normal username = 0");
  assert(scoreUsername("8374629102") > 0, "all digits detected");
  assert(scoreUsername("a".repeat(25)) > 0, "very long username detected");
  assert(scoreUsername("bcdfghjklmnp") > 0, "no vowels detected");
  assert(scoreUsername("user12345678") > 0, "high digit ratio detected");
  assert(scoreUsername("_a_b_c_d_e_") > 0, "underscore heavy detected");
  assert(scoreUsername("_bot_12345678901234567890_bcdfg_") <= 30, "capped at 30");
});

// ── preScoreFromMetadata ────────────────────────────────────────────────────

suite("preScoreFromMetadata", () => {
  const fakeResult = preScoreFromMetadata("bot8374629", 0, false, null, false);
  assert(fakeResult !== null && fakeResult.score >= 75, "obvious fake >= 75");

  const legitResult = preScoreFromMetadata("jean_dupont", 150, false, "Jean Dupont", true);
  assert(legitResult !== null && legitResult.score <= 15, "obvious legit <= 15");

  const ambigResult = preScoreFromMetadata("some_user", 10, false, "Some User", true);
  assert(ambigResult === null, "inconclusive returns null");
});

// ── scoreProfile ────────────────────────────────────────────────────────────

suite("scoreProfile", () => {
  const fakeProfile = {
    username: "bot123456",
    followerCount: 0,
    postCount: 0,
    replyCount: 0,
    isPrivate: false,
    hasBio: false,
    hasProfilePic: false,
    hasLinkInBio: false,
    hasIgLink: false,
    fullName: null,
  };
  const fakeResult = scoreProfile(fakeProfile);
  assert(fakeResult.score >= 50, `fake profile score ${fakeResult.score} >= 50`);

  const legitProfile = {
    username: "jean_dupont",
    followerCount: 200,
    postCount: 50,
    replyCount: 30,
    isPrivate: false,
    hasBio: true,
    hasProfilePic: true,
    hasLinkInBio: true,
    hasIgLink: false,
    fullName: "Jean Dupont",
  };
  const legitResult = scoreProfile(legitProfile);
  assert(legitResult.score < 30, `legit profile score ${legitResult.score} < 30`);
});

// ── Summary ─────────────────────────────────────────────────────────────────

console.log(`\n${"=".repeat(40)}`);
console.log(`Results: ${passed} passed, ${failed} failed`);
if (failed > 0) process.exit(1);
